### DISCUSSION TITLE
Is there a way to preload/cache controls?

### DISCUSSION THREAD

**Discussion by HakanL** [unoplatform/uno#15965] [Q&A]
In my GTK/Skia Uno application running on Raspberry Pi 4 it runs a little sluggish, especially when starting cold. I can see controls being added to the screen step by step (like I'm getting the backgrounds in the grid first, then text blocks (title, etc), then the things in the WrapPanel including icons. Is there a way to programmatically warm it up so the user gets a more responsive UI from cold start?
How would I go about adding logging to determine what's taking time and where it can be optimized? I have looked into AOT, but couldn't get it to work in my environment.

**Comment by jeromelaban** [comment]
Thanks. Preloading is not going to help you here, as it will need to be loaded in all casesregardless of being preloaded.

NativeAOT is likely to be the way to get better performance on cold start.

Can you provide more details about the errors you got?

**Reply by HakanL** [reply]
I'm using EF Core, it seems to be incompatible with AOT in .NET 8. But just trying to publish (to see how far it takes me) gives me this error:
`Cross-OS native compilation is not supported.`
And granted I am ultimately running this on Linux (RPi), but this is just while building on my Win11-PC. Also getting a popup saying a build error with a path to the build log, which has this in it:
```
3/21/2024 2:33:47 PM
System.AggregateException: One or more errors occurred. ---> Microsoft.WebTools.Shared.Exceptions.WebToolsException: Publish has encountered an error. We were unable to determine the cause of the error. Check the output log for more details. 
   --- End of inner exception stack trace ---
---> (Inner Exception #0) Microsoft.WebTools.Shared.Exceptions.WebToolsException: Publish has encountered an error. We were unable to determine the cause of the error. Check the output log for more details. <---

Microsoft.WebTools.Shared.Exceptions.WebToolsException: Publish has encountered an error. We were unable to determine the cause of the error. Check the output log for more details. 

===================
```
And this is with the following added to my head csproj:
```
	<PropertyGroup>
		<PublishAot>true</PublishAot>
	</PropertyGroup>
	<ItemGroup>
		<TrimmerRootAssembly Include="DMXCore100.Skia.Gtk" />
		<TrimmerRootAssembly Include="GtkSharp" />
		<TrimmerRootAssembly Include="GdkSharp" />
	</ItemGroup>
```

FWIW the error I'm getting with EF Core is this: `Model building is not supported when publishing with NativeAOT. Use a compiled model.`. However I am using compiled models, so the error doesn't make sense. But I read that it's a WIP: https://github.com/dotnet/efcore/issues/29754

Not sure if there are any other ways I can optimize performance?

**Reply by jeromelaban** [reply]
You can try to validate if an app without all those dependencies is effectively running faster with AOT and determine if that addresses that cold start issue.

**Reply by HakanL** [reply]
Yeah, that's a good idea, some effort, but I'll check it out.
